### PR TITLE
Adjusted SnippetAdapter to changes in Umberto

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Internal: Adjusted `SnippetAdapter` to changes in Umberto. Improved the `whitelisted-snippet` option that has been renamed to `--snippets`.

Now it understands glob patterns. `--snippets=examples/**,` will build all snippets that starting with `examples/`. You can specify more than single glob pattern: `--snippets=framework/tutorials/*, examples/bootstrap-ui`

Do not merge it if Umberto has not been released.

### CODE REVIEW

File: scripts/docs/build-docs.js [hunk 1]
```diff
@@ -53,11 +53,11 @@ function runUmberto( options ) {
 		skipLiveSnippets: options.skipLiveSnippets,
 		skipValidation: options.skipValidation,
 		snippetOptions: {
-			production: options.production
+			production: options.production,
+			whitelistedSnippets: whitelistedSnippets ? whitelistedSnippets.replace( '--whitelisted-snippet=', '' ).split( ',' ) : []
```
Comment (msamsel) [inline scripts/docs/build-docs.js]: Maybe '--whitelisted-snippet=' should be now rename to '--whitelisted-snippets=' as this option on umberto side now supports array or globs.

Comment (Reinmar) [reply scripts/docs/build-docs.js]: Can it be simply `--snippets`? We have `--skip-snippets` already. Typing `--whitelisted-snippets` is time-consuming.

Comment (pomek) [reply scripts/docs/build-docs.js]: `--snippets` sounds good.

File: scripts/docs/snippetadapter.js [hunk 2]
```diff
@@ -7,83 +7,242 @@
...
+const DEFAULT_LANGUAGE = 'en';
 
-module.exports = function snippetAdapter( data ) {
-	if ( !data.snippetSource.js ) {
-		throw new Error( `Missing snippet source for "${ data.snippetPath }".` );
+/**
+ * @param {Set.<Snippet>} snippets Snippet collection extracted from documentation files.
+ * @param {Object} options
+ * @param {Boolean} options.production Whether to build snippets in production mode.
+ * @param {Array.<String>>} options.whitelistedSnippets An array that contains glob patterns
```
Comment (pomek) [inline scripts/docs/snippetadapter.js]: Missing a dot at the end of the sentence.

File: scripts/docs/snippetadapter.js [hunk 3]
```diff
@@ -7,83 +7,242 @@
...
 
-function getWebpackConfig( config ) {
+/**
+ * Removes snippets that names do not match to patterns specified in `whitelistedSnippets` array.
+ *
+ * @param {Set.<Snippet>} snippets Snippet collection extracted from documentation files.
+ * @param {Array.<String>} whitelistedSnippets Snippet patterns that should be built.
+ */
+function filterWhitelistedSnippets( snippets, whitelistedSnippets ) {
+	if ( !whitelistedSnippets.length ) {
```
Comment (msamsel) [inline scripts/docs/snippetadapter.js]: whitelistedSnippets might be not defined ( this happen in big bang docs ), you should handle such case, and build all snippets in such scenario.
```
Executing the snippet adapter for "ckeditor5".
TypeError: Cannot read property 'length' of undefined
    at filterWhitelistedSnippets (/bigbang-docs.ckeditor.com/projects/ckeditor5/scripts/docs/snippetadapter.js:200:28)
    at snippetAdapter (/bigbang-docs.ckeditor.com/projects/ckeditor5/scripts/docs/snippetadapter.js:78:2)
    at promise.then (/umberto/scripts/filter/after-post-render/snippets.js:48:12)
```

**Comment by oleq** [comment]
I just wanted to leave a note that this would help me a lot. We have guides that build the editor in one snippet (to avoid core duplication) and examples in other snippets. So to build a particular snippet you have to build two pieces like this:

```
yarn docs --whitelisted-snippet=build-mention-source,mention-customization --skip-api --skip-validation
```

which is impossible ATM.

**Comment by pomek** [comment]
> which is impossible ATM

After closing this PR, it will work. You will be able to use a glob pattern. See the merge message.

File: scripts/docs/snippetadapter.js (hunk 3 above)
Comment (pomek) [reply scripts/docs/snippetadapter.js]: Fixed.

Comment (pomek) [reply scripts/docs/snippetadapter.js]: Now the function will not be called if `whitelistedSnippets` is `undefined`.

File: scripts/docs/build-docs.js (hunk 1 above)
Comment (pomek) [reply scripts/docs/build-docs.js]: Renamed to `--snippets` for CLI.

File: scripts/docs/snippetadapter.js [hunk 4]
```diff
@@ -7,83 +7,244 @@
...
 
-	const webpackConfig = getWebpackConfig( {
-		entry: data.snippetSource.js,
-		outputPath,
-		language: snippetConfig.language,
-		production: data.options.production,
-		definitions: data.options.definitions || {}
-	} );
+	// Remove snippets that do not match to patterns specified in `options.whitelistedSnippets`.
+	if ( options.whitelistedSnippets ) {
```
Comment (ma2ciek) [inline scripts/docs/snippetadapter.js]: Now it's present all the time.

File: scripts/docs/snippetadapter.js [hunk 5]
```diff
@@ -7,83 +7,244 @@
...
-	} else {
-		promise = runWebpack( webpackConfig );
-		webpackProcesses.set( outputPath, promise );
+	// Nothing to build.
+	if ( !webpackConfigs.length ) {
+		return promise;
+	}
+
+	for ( const config of webpackConfigs ) {
+		promise = promise.then( () => runWebpack( config ) );
```
Comment (ma2ciek) [inline scripts/docs/snippetadapter.js]: I'd check the `Promise.all()` too, but that's fine too.

File: scripts/docs/snippetadapter.js (hunk 4 above)
Comment (ma2ciek) [reply scripts/docs/snippetadapter.js]: Or I might be wrong as this file will be executed by the bigbangdocs...

File: scripts/docs/snippetadapter.js (hunk 5 above)
Comment (pomek) [reply scripts/docs/snippetadapter.js]: `Promise.all()` = 2 webpacks at the same time = (most probably) memory errors.

File: scripts/docs/snippetadapter.js (hunk 4 above)
Comment (mlewand) [reply scripts/docs/snippetadapter.js]: In that case I'm marking this as resolved.

**Comment by Reinmar** [comment]
Works like a charm and it's much faster. Great job!

Waiting with merging this for the new umberto version.

**Pull request merged**